### ISSUE TITLE
Java Enum.valueOf broken in current code

### ISSUE THREAD

**Issue by harrycheung** [google/j2objc#458]
995ca26 "functinalizes" method calls too aggressively. Before the commit, the translated Enum code would use Objective-C message syntax, but after the commit, the code would use C function syntax. By using the C function syntax, the Enum's initializer never gets called causing the Java translated method <b>.valueOf</b> to fail.

In Java:
```
MyEnum.valueOf("B")
```
Translated before commit:
```
[MyEnumEnum valueOfWithNSString:@"B"]
```
Translated after commit (not working):
```
MyEnumEnum_valueOfWithNSString_(@"B")
```

Here's how you replicate: https://github.com/harrycheung/testJ2Enum

**Comment by foens** [comment]
I just got knocked down by this issue as well. What is the workaround more precisely, what code to add and exactly where?

**Comment by tomball** [comment]
Here's the patch:

==== /a/j2objc/translator/src/main/java/com/google/devtools/j2objc/translate/EnumRewriter.java#10 - /b/j2objc/translator/src/main/java/com/google/devtools/j2objc/translate/EnumRewriter.java#11 ====
@@ -135,12 +135,13 @@
     StringBuilder sb = new StringBuilder();
     sb.append(String.format(
         "FOUNDATION_EXPORT IOSObjectArray *%s_values() {\n"
+        + "  %s_init();\n"
         + "  return [IOSObjectArray arrayWithObjects:%s_values_ count:%s type:"
         + "[IOSClass classWithClass:[%s class]]];\n"
         + "}\n"
         + "+ (IOSObjectArray *)values {\n"
         + "  return %s_values();\n"
-        + "}\n\n", typeName, typeName, numConstants, typeName, typeName));
+        + "}\n\n", typeName, typeName, typeName, numConstants, typeName, typeName));
 
     sb.append(String.format(
         "+ (%s *)valueOfWithNSString:(NSString *)name {\n"
@@ -149,12 +150,13 @@
 
     sb.append(String.format(
         "%s *%s_valueOfWithNSString_(NSString *name) {\n"
+            + "  %s_init();\n"
             + "  for (int i = 0; i < %s; i++) {\n"
             + "    %s *e = %s_values_[i];\n"
             + "    if ([name isEqual:[e name]]) {\n"
             + "      return e;\n"
             + "    }\n"
-            + "  }\n", typeName, typeName, numConstants, typeName, typeName));
+            + "  }\n", typeName, typeName, typeName, numConstants, typeName, typeName));
     if (Options.useReferenceCounting()) {
       sb.append(
           "  @throw [[[JavaLangIllegalArgumentException alloc] initWithNSString:name]"

**Comment by tomball** [comment]
Here it is, without Markdown trying to format it:
```objc
==== //depot/google3/third_party/java_src/j2objc/translator/src/main/java/com/google/devtools/j2objc/translate/EnumRewriter.java#10 - //depot/google3/third_party/java_src/j2objc/translator/src/main/java/com/google/devtools/j2objc/translate/EnumRewriter.java#11 ====
# action=edit type=text
--- google3/third_party/java_src/j2objc/translator/src/main/java/com/google/devtools/j2objc/translate/EnumRewriter.java	2014-10-08 05:39:33.000000000 -0700
+++ google3/third_party/java_src/j2objc/translator/src/main/java/com/google/devtools/j2objc/translate/EnumRewriter.java	2015-01-05 11:31:45.000000000 -0800
@@ -135,12 +135,13 @@
     StringBuilder sb = new StringBuilder();
     sb.append(String.format(
         "FOUNDATION_EXPORT IOSObjectArray *%s_values() {\n"
+        + "  %s_init();\n"
         + "  return [IOSObjectArray arrayWithObjects:%s_values_ count:%s type:"
         + "[IOSClass classWithClass:[%s class]]];\n"
         + "}\n"
         + "+ (IOSObjectArray *)values {\n"
         + "  return %s_values();\n"
-        + "}\n\n", typeName, typeName, numConstants, typeName, typeName));
+        + "}\n\n", typeName, typeName, typeName, numConstants, typeName, typeName));
 
     sb.append(String.format(
         "+ (%s *)valueOfWithNSString:(NSString *)name {\n"
@@ -149,12 +150,13 @@
 
     sb.append(String.format(
         "%s *%s_valueOfWithNSString_(NSString *name) {\n"
+            + "  %s_init();\n"
             + "  for (int i = 0; i < %s; i++) {\n"
             + "    %s *e = %s_values_[i];\n"
             + "    if ([name isEqual:[e name]]) {\n"
             + "      return e;\n"
             + "    }\n"
-            + "  }\n", typeName, typeName, numConstants, typeName, typeName));
+            + "  }\n", typeName, typeName, typeName, numConstants, typeName, typeName));
     if (Options.useReferenceCounting()) {
       sb.append(
           "  @throw [[[JavaLangIllegalArgumentException alloc] initWithNSString:name]"
```

**Comment by tomball** [comment]
Fixed in 0.9.6.
